App Settings Cleanup, Part 1

mindgraffiti opened the pull request:
Ref #11925

## Summary
Part 1 of this PR updates some of the UI in App Settings to follow our current color scheme and improves the UX for clearing the Media Cache. This PR does not address the changes to the "Max Image Upload Size" row or the new location option. It is arriving in a separate future PR.

**Changes include:**
- Clear Spotlight Index button uses blue brand color and is left aligned
- Clear Siri Shortcut Suggestions uses blue brand color and is left aligned
- Media Cache Size UI is updated to show a disclosure arrow and navigates to its own ViewController 
- "Other" section header remains intact because more options are listed in that section compared to when the ticket was created

## Screenshots
| Before | After |
| --- | --- |
| ![Simulator Screen Shot - iPhone 12 Pro - 2022-05-04 at 14 20 06](https://user-images.githubusercontent.com/1062444/167921306-c5ee76b8-8add-41a0-a376-77ed74e90ee5.png) | ![Simulator Screen Shot - iPhone 12 Pro - 2022-05-04 at 16 04 50](https://user-images.githubusercontent.com/1062444/167921360-74a9e31e-48ce-42f1-93ef-0a7c48ee1fae.png) |

## To test:
0. Install the app on a device, with iOS 12 or higher. Steps 7 and 8 are features found only in iOS 12 and up.
1. Select "My Site" in the bottom nav baar
2. Select your profile picture
3. Choose "App Settings" in the modal view
4. The "Media Cache Size" row should display the current size of the media cache
5. Select the row. It should navigate to a new ViewController with the "Media Cache Size" title and size row, as well as a row button to clear the cache. Clear the cache and verify the size reports "Empty"
6. Navigate back to App Settings
7. Tap the "Clear Spotlight Index" and wait for a toast message that it was successful
8. Tap the "Clear Siri Shortcut Suggestions" and wait for a toast message that it was successful 

## Regression Notes
1. Potential unintended areas of impact
- App Settings
- WPImmuTableRows contains a new row button style

6. What I did to test those areas of impact (or what existing automated tests I relied on)
Run all unit tests, which includes:
- ImmuTableTest 
- MediaFileManagerTests 

10. What automated tests I added (or what prevented me from doing so)
- None, UI tests don't seem like a valuable use of time

PR submission checklist:
- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

mindgraffiti committed:
Create a custom App Settings row

Must be left aligned and blue text

mindgraffiti committed:
Add BrandedButtonRow to WPImmuTableRows

This row is a text-only row button, left aligned, using the WordPress blue brand color for the text

mindgraffiti committed:
Replace custom row in ViewController with reusable BrandedButtonRow

mindgraffiti committed:
Add MediaCacheSettingsViewController to Me > App Settings

mindgraffiti committed:
Import methods from AppSettingsViewController wholesale

Add the table, rows, actions, and other functions to make the Media Cache View Controller work like the new design.

mindgraffiti committed:
Change the MediaCacheRow type and navigate to new ViewController

Navigate to the MediaCacheSettingsViewController upon tapping the MediaCacheRow. Remove the mediaClearCacheRow from the AppSettingsViewController. It has moved to the MediaCacheSettingsViewController.

mindgraffiti committed:
Merge branch 'trunk' into feature/11925-app-settings-cleanup

mindgraffiti committed:
Update the Media Cache table to use plain styling, not grouped

mindgraffiti committed:
Remove unused function

mindgraffiti committed:
Merge branch 'trunk' into feature/11925-app-settings-cleanup

mindgraffiti committed:
Add a line to the release notes describing minor changes to App Settings

Gio2018 commented on WordPress/Classes/Utility/WPImmuTableRows.swift:
```diff
@@ -236,6 +236,34 @@ struct LinkWithValueRow: ImmuTableRow {
 
+/// Create a row button: uses the WordPress branded blue and is left aligned
+///
+struct BrandedButtonRow: ImmuTableRow {
```
I know it reflects pre-existing code, but maybe we can give this type a name that clarifies a bit better what it is. The cell does not actually contain a button, and it's specifically used in App Settings, so Maybe we can use something like `SettingsActionRow` or similar? What do you think?

Gio2018 commented on WordPress/Classes/Utility/WPImmuTableRows.swift:
```diff
@@ -236,6 +236,34 @@ struct LinkWithValueRow: ImmuTableRow {
 
+/// Create a row button: uses the WordPress branded blue and is left aligned
+///
+struct BrandedButtonRow: ImmuTableRow {
```
Maybe we can give this type a name that clarifies a bit better what it is. The cell does not actually contain a button, and it's specifically used in App Settings, so Maybe we can use something like `SettingsActionRow` or similar? What do you think?

mindgraffiti replied:
Agreed. `BrandedNavigationRow` maybe? This won't only apply to the Settings screen. I'd like to replace the "Help and Support" ViewController row stylings with the ImmuTableRow styles to cut down on repetitive code.

Gio2018 commented on WordPress/Classes/Utility/WPImmuTableRows.swift:
```diff
@@ -236,6 +236,34 @@ struct LinkWithValueRow: ImmuTableRow {
+    }
+
+    func configureCell(_ cell: UITableViewCell) {
+        let cell = cell as! WPTableViewCellIndicator
```
since this is an internal method that can be called from anywhere in the code, we do not seem to have a full guarantee that the passed type will actually be `WPTableViewCellIndicator`, so I think we should remove the force unwrap in favor of a `guard` statement.

mindgraffiti replied:
Oh gosh. I've already forgotten about the traps that force unwrapping causes. Thank you, will replace.

Gio2018 commented on WordPress/Classes/Utility/WPImmuTableRows.swift:
```diff
@@ -236,6 +236,34 @@ struct LinkWithValueRow: ImmuTableRow {
+    }
+
+    func configureCell(_ cell: UITableViewCell) {
+        let cell = cell as! WPTableViewCellIndicator
```
since this is an internal method that can be called from anywhere in the code, we do not seem to have a full guarantee that the passed type will actually be `WPTableViewCellIndicator`, so I think we should remove the force unwrap in favor of a `guard` statement.

Gio2018 commented on WordPress/Classes/ViewRelated/Me/App Settings/MediaCacheSettingsViewController.swift:
```diff
@@ -0,0 +1,124 @@
+import Foundation
+
+class MediaCacheSettingsViewController: UITableViewController {
+    fileprivate var handler: ImmuTableViewHandler!
```
I know this descends from how `ImmutableViewHandler` is structured, but maybe we can use a regular optional instead of implicitly unwrap, and have the assignment `handler = ImmuTableViewHandler(takeOver: self)` moved over in the initializer, right after the `super.init(_:)` call?

Gio2018 replied:
I know this descends from how `ImmutableViewHandler` is structured, but maybe we can use a regular optional instead of implicitly unwrap, and have the assignment `handler = ImmuTableViewHandler(takeOver: self)` moved over in the initializer, right after the `super.init(_:)` call?

Gio2018 reviewed (changes requested):
Thanks for putting this together @mindgraffiti !

It looks mostly good to me; I left a few code comments for you to consider.
Other than that, I am not totally sure if we should leave the "Clear Device Media Cache" in the main screen, as it could not be immediately obvious to users that they can do so by navigating to the "Media Cache Size" screen, but I'll let @iamthomasbishop and @ScoutHarris  chime in for that!

mindgraffiti committed:
Merge branch 'trunk' into feature/11925-app-settings-cleanup

mindgraffiti committed:
Rename new ImmuTableRow to more descriptive name

From BrandedButtonRow (it's not a button) to BrandedNavigationRow (it does navigate to a new ViewController)

mindgraffiti committed:
Remove force unwrap

iamthomasbishop reviewed (changes requested):
@mindgraffiti  Nice work here, overall a lot cleaner! A couple of things jumped out at me, but nothing huge.

I just saw @Gio2018's comment about moving the "Clear device media cache" setting to a sub-page and I don't have a huge problem with it being down a level because I'm not sure if I'd consider it a "top-level" app setting. However, I think we should consider changing the label to be a bit more generic, i.e. "Media Cache". If I were on the hunt for a place to clear my cache, I think that'd be clearer than "Media Cache *Size*".

I'm not sure if this is within the scope of your work here, but it'd be nice to consolidate the "Max Image Upload Size" top-level settings in a  similar way to how you've handled the "Media Cache" settings. Something along the lines of this, perhaps — thoughts?

<a href="https://user-images.githubusercontent.com/1200432/168128793-9a960b32-ccec-4133-8560-65c7fe214ff6.png"><img src="https://user-images.githubusercontent.com/1200432/168128793-9a960b32-ccec-4133-8560-65c7fe214ff6.png" width="800"></a>


A couple of suggestions for the "Media Cache" sub-page:

- I think it'd make sense to split apart the "Media Cache Size" row and "Clear Device Media Cache" button row to provide some separation, considering they are separate functions
- I think we may want to use the inset table row style here unless I'm missing something

Here's a visual of what I'm thinking with these suggestions:

<a href="https://user-images.githubusercontent.com/1200432/168128262-cc77a6cc-49e1-4e9a-88f8-2486094474f5.png"><img src="https://user-images.githubusercontent.com/1200432/168128262-cc77a6cc-49e1-4e9a-88f8-2486094474f5.png" width="800"></a>

mindgraffiti committed:
Change label from "Media Cache Size" to "Media Cache"

Improves the user's understanding for this row

mindgraffiti committed:
Remove force unwraps, move handler takeover to init method

mindgraffiti committed:
Make table inset

mindgraffiti committed:
Change Media Cache rows to be separated instead of grouped

mindgraffiti commented:
@Gio2018 @iamthomasbishop changes made. Ready for a second review!

- [x] "Media Cache" instead of "Media Cache Size"
- [x] Max Image Upload ... understood. This will be worked on in a separate PR
- [x] Inset table and add spacing between the two rows on the Media Cache ViewController:

<img src="https://user-images.githubusercontent.com/1062444/168140977-a0df1a53-4c9b-4d71-8bd1-d8dc0d30f5e4.png" width="300" />

mindgraffiti committed:
Merge branch 'trunk' into feature/11925-app-settings-cleanup

Gio2018 commented:
>  I'm not sure if I'd consider it a "top-level" app setting.

makes a lot of sense @iamthomasbishop ! I also agree that just changing to "Media Cache" as @mindgraffiti has already done) would be the best option here.

mindgraffiti committed:
Merge branch 'trunk' into feature/11925-app-settings-cleanup

mindgraffiti committed:
Merge branch 'trunk' into feature/11925-app-settings-cleanup

mindgraffiti committed:
Revert "Merge branch 'trunk' into feature/11925-app-settings-cleanup"

This reverts commit 1e09f4e38851f481fce4cf860ec91aa34ce52472.

Gio2018 reviewed (changes requested):
hey @mindgraffiti , it looks like you need to merge trunk into this branch, it looks like it's based on an older commit now?

mindgraffiti committed:
Merge branch 'trunk' into feature/11925-app-settings-cleanup

mindgraffiti commented:
@Gio2018 trunk was breaking the ability to build, so I reverted. It should be good now.

mindgraffiti committed:
Moving PR to 20.0 release

mindgraffiti committed:
Merge branch 'trunk' into feature/11925-app-settings-cleanup

hassaanelgarem committed:
Merge pull request #18591 from wordpress-mobile/issue/18563-qs-default-section-update

Quick Start: Refresh Quick Start card location when the default section changes

hassaanelgarem committed:
Revert "Merge pull request #18591 from wordpress-mobile/issue/18563-qs-default-section-update"

This reverts commit 47d49cdcd8beda697efec7bbc5a326f34fe2fc40.

mindgraffiti committed:
Revert "Revert "Merge branch 'trunk' into feature/11925-app-settings-cleanup""

This reverts commit f5f2b3cfd803666286b74185e7a914274248211d.

mindgraffiti committed:
Merge branch 'trunk' into feature/11925-app-settings-cleanup

mindgraffiti committed:
Revert "Merge branch 'trunk' into feature/11925-app-settings-cleanup"

This reverts commit 92eed6be79b80f800139ea6173a3fd3314b03549.

mindgraffiti committed:
Update release notes

mindgraffiti committed:
Revert "Update release notes"

This reverts commit 99a4573aeb5846a0cf02a6d7b5c58252f6b91ac9.

mindgraffiti committed:
Merge branch 'trunk' into feature/11925-app-settings-cleanup

mindgraffiti committed:
Merge branch 'trunk' into feature/11925-app-settings-cleanup

mindgraffiti commented:
Closing in favor of PR #18636
